File: stacperf/ruby/stacperf.rb
#!/usr/bin/ruby

class Stacperf
	MEMSIZE=[16, 32, 64, 128, 256, 512, 1024,
		2048, 4096, 8192, 16384, 32768, 65536,
		131072, 262144, 524288, 1048576,
		2097152, 4194304, 8388608, 10000000 ]
	def run
		a=Array.new(MEMSIZE[-1],0.0)

		print "N,\tseconds\t,Mop/s,\tchecksum\n"
		MEMSIZE.each { |n|
			print n.to_s() + ",\t"
			checksum=0.0
			nloops=(MEMSIZE[-1]/n).to_i()

			start_time=Time.now.to_f()
			
			nloops.times {
				(0..n-1).each { |i|
					a[i]=i.to_f()
				}
				(0..n-1).reverse_each { |i|
					checksum += a[i]
				}
			}

			end_time=Time.now.to_f()
			elapsed_time=end_time-start_time
			print elapsed_time.to_s()+",\t"
			
			mops=0.000001*2.0*n*nloops/elapsed_time
			print mops.to_s()+",\t"+checksum.to_s()+"\n"
		}
	end
end

sp=Stacperf.new()
sp.run()

File: stacperf/c/stacperf.c
/* pesudo-stack operation performance */
#include <stdlib.h>
#include <stdio.h>
#include <sys/time.h>
#include <linux/compiler.h>

static double dtime(void);
extern int main(void);

static double dtime()
{
	double q;
	struct timeval tnow;

	gettimeofday(&tnow,NULL);
	q = (double)tnow.tv_sec + (double)tnow.tv_usec * 1.0e-6;

	return q;
}

int main() {
	static long memsize[] = {
		16l, 32l, 64l, 128l, 256l, 512l, 1024l,
		2048l, 4096l, 8192l, 16384l, 32768l, 65536l,
		131072l, 262144l, 524288l, 1048576l,
		2097152l, 4194304l, 8388608l, 10000000l};
	long i;
	long max_n=memsize[sizeof(memsize)/sizeof(memsize[0])-1];
	long *page_a;
	if(!(page_a=(long*)valloc(sizeof(long)*max_n))) {
		puts("memory allocation failed");
		exit(EXIT_FAILURE);
	}

	/* initial initialization array, own address */
	for(i=0; i<max_n; i++) { page_a[i]=i; }
	fputs("array init done\n",stderr);

	puts("N,\tseconds,\tMop/s,\tchecksum");
	for(i=0; i<sizeof(memsize)/sizeof(memsize[0]); i++) {
		long n=memsize[i];
		long nloops=10L*max_n/n;
		double start_time,end_time;
		long chksum=0l;
		double elapsed_time, mops;
		long i;
		printf("%ld,\t",n); fflush(stdout);
		// OSMemoryBarrier();
		start_time=dtime();
		// OSMemoryBarrier();

		i=nloops;
		do {
			long j;
			for(j=0; j<n; j++) { page_a[j]=j; }
			for(j=n-1; j>=0; j--) { chksum += page_a[j]; }
		} while(--i);

		// OSMemoryBarrier();
		end_time=dtime();
		
		elapsed_time=end_time-start_time;
		mops=1.E-6*2.*((double)(n*nloops))/elapsed_time;
		printf("%f,\t%f,\t%016lX\n",elapsed_time,mops,chksum);
	}

	free(page_a);
	exit(EXIT_SUCCESS);
}

